### PR TITLE
buildTime and serverStartupTime substitutions for page headers

### PR DESCRIPTION
**Pull request by labkey-jeckels**
#### Rationale
Some new properties that admins can use for a dynamic site short name: `buildTime` (from the Core module) and `serverStartTime`

#### Changes
* Add the new properties
* Don't return null (yielding NPEs) if there's an unrecognized substitution in the header expression
* New StringExpression option to pass unrecognized substitution properties through as-is
* Minor code cleanup

### CODE REVIEW

File: api/src/org/labkey/api/util/StringExpressionFactory.java [hunk 1]
```diff
@@ -693,6 +729,12 @@ public static AbstractStringExpression.NullValueBehavior fromXML(StringExpressio
                 return null;
             }
 
+            public abstract String handleNull(StringExpressionFactory.StringPart part) throws StopIteratingException;
+
+            public String handleUndefined(StringExpressionFactory.StringPart part) throws StopIteratingException
```
Comment (labkey-adam) [inline api/src/org/labkey/api/util/StringExpressionFactory.java:734]: Thanks for pushing these into the enum!

**Comment by labkey-adam** [comment]
Would be nice to display the new properties on the Server Information page, in export diagnostics, etc. for consistency

**Review by labkey-adam — APPROVED** [review]

**Comment by labkey-jeckels** [comment]
> Would be nice to display the new properties on the Server Information page, in export diagnostics, etc. for consistency

I added them to Server Info and exported diagnostics. I couldn't find any other spots that use the existing properties, so let me know if I missed anything to cover the "etc"

**Comment by labkey-adam** [comment]
> > Would be nice to display the new properties on the Server Information page, in export diagnostics, etc. for consistency
> 
> I added them to Server Info and exported diagnostics. I couldn't find any other spots that use the existing properties, so let me know if I missed anything to cover the "etc"

Looks good... the other usages I was remembering appear to be covered by reflection-based map creation

**Pull request merged**